### DISCUSSION TITLE
raspberry pi 2b+ being treated like a pi zero.

### DISCUSSION THREAD

**Discussion by dosuserx** [probonopd/MiniDexed#550] [Q&A]
hey I haven't found an answer to this yet so I'm asking on here. i have a pi in all the major versions and ive set up minidexed on all of them to gauge quality and one thing I've noticed is that my raspberry pi 2b+ is not showing 8 tone generators, it's only showing 1. i haven't found any reason for this and i searched on here. 

im using the same sd card and hardware, just swaping the pis out. I'm gonna go back and grab my pi 3 when i get home but my 2 pi 4s show 8 TGs. is there a setting so i can force it and see if it works? my pi 2b+ is the one i was hoping to install in my project.

**Comment by Banana71** [comment]
Hello Dos user,
This is a new problem for me.
Only when you set the reverb time to high values and the compressor was switched on did my RPi 2b+ sometimes drop out a few times, but otherwise I see all 8 TGs.
Which version do you have installed on your SD card?

**Reply by dosuserx** [reply]
good question, cuz i feel like i didn't have this problem last month. i actually don't know if that's written anywhere but the files were modified on sept 23, so there's a good bet that its the 9-23 release. i may go back and try the previous release and ill let you know if that fixes it.

**Reply by Banana71** [reply]
And I'll try out the version from 23.09. on my RPi 2b+ later and report back.

**Reply by Banana71** [reply]
My RPi 2 works with the version from September 23, 2023.
Which Raspberry 2 do you have exactly?

<html>
<body>
<h4 class="wp-block-heading" style="box-sizing: border-box; font-family: Roboto, sans-serif; color: var(--td_text_color, #111111); font-weight: 400; margin: 24px auto 14px; font-size: 19px; line-height: 29px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span id="Raspberry_Pi_Baureihe_2" style="box-sizing: border-box;">Raspberry Pi Baureihe 2</span></h4><figure class="wp-block-table" style="margin: 0px auto; box-sizing: border-box; display: block; overflow-x: auto; color: rgb(34, 34, 34); font-family: open-sans-v15-latin-regular; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Modell | Pi 2 Modell B | Pi 2 Modell B v1.2
-- | -- | --
Veröffentlichung | Februar 2015 | September 2016
Größe der Platine mm | 85,6×56 | 85,6×56
CPU | Cortex-A7 | Cortex-A53
Kerne | 4 | 4
RAM | 1 GB | 1 GB
Anschlüsse | 3,5″ Klinke für Composite-Video und AudioHDMI4x USB 2.01x Micro-USB zur Stromversorgung | 3,5″ Klinke für Composite-Video und AudioHDMI4x USB 2.01x Micro-USB zur Stromversorgung
Netzwerk | 10/100 Mbit/s Ethernet | 10/100 Mbit/s Ethernet
Stromaufnahme | 800 mA | 800 mA

</figure>
</body>
</html>

**Reply by dosuserx** [reply]
yes. my pi 2 is a raspberry pi 2b+ version 1.2. its also dated 2014.

also i confirmed that both the current release and the previous release (august) have the same effect. this same card works correctly on my pi 3b v1.2, my new/used pi4 2gig and my pi4 4gig.

im stumped.

**Reply by Banana71** [reply]
Then the RPi 2 V1.2 will probably not be recognized as a multicore CPU. Can anyone confirm this? Unfortunately I don't have Rpi 2 V1.2 to test it myself. Can the problem be solved in config.h from line 34?

config.h 
 

> #ifndef ARM_ALLOW_MULTI_CORE
        static const unsigned ToneGenerators = 1;
#else
        static const unsigned TGsCore1 = 2;             // process 2 TGs on core 1
        static const unsigned TGsCore23 = 3;            // process 3 TGs on core 2 and 3 each
        static const unsigned ToneGenerators = TGsCore1 + 2*TGsCore23;
#endif

**Reply by diyelectromusic** [reply]
ARM_ALLOW_MULTI_CORE is defined for RPI>1 in build.sh so it should be ok.  As the release build contains a range of kernel images I wonder if it is somehow picking kernel.img rather than kernel7.img...?
 
 Is it worth trying to delete the other kernel images to see what happens?
 
 The different kernel instances to Cortex architectures used in the build can be found here: https://github.com/rsta2/circle#building
 
 Kevin

**Reply by diyelectromusic** [reply]
Ok, so I don't know what the 2B+ should appear as, as it has the same Cortex-A53 as the Pi 3... so I wonder if that would need the kernel8-32.img to work, which I don't think we build?

We build a 32-bit kernel for RPI 1 and 2  (kernel.img and kernel7.img) and 64-bit for 3 and 4 (kernel8.img and kernel8-rpi4.img).

I guess it might work with kernel8.img?  But our config.txt file only selects arm_64bit=1 for the 3 and 4...

Rather unhelpfully it appears that config.txt syntax states the following:
[pi2] = Model 2B (BCM2836- or BCM2837-based)

So doesn't seem to be able to distinguish between a A7 or A53 based CPU for the Pi 2...

So yes, I think config.txt tells it a Pi 2 is 32 bit, but there is no 32-bit A53 kernel, so I guess it must fall back on the original kernel.img, which isn't multicore...?

Try adding the following to config.txt as a test - although we can't rely on this as the original Pi 2 would break!

[pi2]
arm_64bit=1

Or if that doesn't work, we might have to specify the kernel directly (circle says some versions of the Pi2 can run with a RPI=3 kernel!)

[pi2]
kernel=kernel8.img

Kevin

**Reply by dosuserx** [reply]
ok so with the pi 2 config.txt, i have tried the following without any success or change:

1)
```
[pi2]
arm_64bit=1

```

2)
```
[pi2]
kernel=kernel8.img
```

3)
```
[pi2]
arm_64bit=1
kernel=kernel8.img
```
 thanks for the help everybody but i think minidexed is just not compatible. i noticed in the code it explicitly doesn't mention the 2nd model, there's no reason in for the program to recognize the pi 2b+ as anything but a model a. maybe it could be worked on by smarter people than myself, in the mean time i guess i'll install my pi3 into the project instead. 

thanks for the help everybody. i really appreciate it.

**Reply by diyelectromusic** [reply]
Unfortunately I only have a Pi 2 1.1 so can't try it, but the things I've read basically say it ought to work like a Pi 3.

What are the symptoms of the failure?  Your original posts said it boots but there is only one TG.  With the above changes and only having the file kernel8.img, does it actually boot at all?  Because if it does, then the code in kernel8.img only knows about 8 TGs! :)

Kevin

**Reply by dosuserx** [reply]
the main symptoms are that the pi 2 shows only 1 TG when i start it up. all the other pis i have used work as they should. but the pi 2 v1.2 will work exactly as if it is a pi zero so it's not total failure, id just rather give up my pi 2 than my 3 but it's not a huge deal.

**Reply by diyelectromusic** [reply]
I know that was your original issue, so if it is still doing that with the above config changes then it is still booting from kernel.img and not the newer ones.  Can you try removing all kernel images apart from kernel8.img and let us know what it does with the various config.txt variations (the original, then the changes suggested above)?  As I say, I don't have this variant myself so can't actually test it for you to tell you what to do.

If we can get it to actually boot kernel8.img then it should be fine as CPU and IO wise it is effectively a slightly slower Pi 3.  Quite why they released it that way I have no idea...

Kevin

**Reply by diyelectromusic** [reply]
Another option, try forcing the use of kernel7.img (32-bit, quad core, 8 TG version) as follows:

[pi2]
kernel=kernel7.img

Kevin

**Reply by dosuserx** [reply]
cool, I'll try it out this weekend and let you know how it goes.

**Reply by probonopd** [reply]
I remember having tested it on a Pi 2 a while ago and it showed all TGs. I think I had probably tested the older of the two versions (wasn't even aware there was a newer Pi 2 iteration). So, if all else fails, maybe go back to a (very) old version of MiniDexed and see if it worked there.
